csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs

```diff
@@ -14,18 +14,65 @@
 // limitations under the License.
 
 using System;
+using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
 
 namespace Apache.Arrow
 {
     public static class SpanExtensions
     {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static Span<T> CastTo<T>(this Span<byte> span)
             where T: struct =>
             MemoryMarshal.Cast<byte, T>(span);
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static ReadOnlySpan<T> CastTo<T>(this ReadOnlySpan<byte> span)
             where T: struct =>
                 MemoryMarshal.Cast<byte, T>(span);
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void CopyToFix<T>(this ReadOnlySpan<T> source, Span<T> target)
+        {
+            CopyToFix(source, 0, target, 0, source.Length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void CopyToFix<T>(this ReadOnlySpan<T> source, T[] target)
+        {
+            CopyToFix(source, 0, target, 0, source.Length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void CopyToFix<T>(this Span<T> source, Span<T> target)
+        {
+            CopyToFix(source, 0, target, 0, source.Length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void CopyToFix<T>(this Memory<T> source, Memory<T> target)
+        {
+            CopyToFix(source.Span, 0, target.Span, 0, source.Length);
+        }
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void CopyToFix<T>(ReadOnlySpan<T> source, int sourceOffset, Span<T> target, int targetOffset, int length)
+        {
+            for (int i = 0; i < length; ++i)
```

Comment:
This is a slower way of copying, and the underlying bug doesn't affect .NET Core.

We already multi-target for .NET Core. So can we only do the slow/workaround way when targeting `netstandard1.3`? And when we are on .NET Core, we can continue to use `Span.CopyTo`, which will perform faster than doing it one element at a time.

See https://github.com/apache/arrow/blob/e24825ac301d19d0b5f9f1dcdc404af98665a24f/csharp/src/Apache.Arrow/Apache.Arrow.csproj#L35-L40 for how we can switch out the implementation for different TFMs. You should be able to do the same thing for `SpanExtensions.cs`.

Reply:
@eerhardt  -  targeting `netcoreapp2.1` makes no sense (in my opinion)  it should just be various netstandardX.Ys and Net Framworks if it were to be multi-target

I did mention in the Jira I would make it multi-target, but the above code makes me think the multi-targeting is not setup correctly

Reply:
We multi-target to `netcoreapp2.1` so we can take advantage of newer APIs that don't exist in netstandard. That way when this library is used in an application running on .NET Core, it can use the newer APIs.

Your new code is a perfect example. Using the built-in `CopyTo` method is faster because it eventually calls `memcpy`, which will copy data faster than doing it one element at a time. Since there is no issue using the built-in `CopyTo` on .NET Core, it doesn't make sense to take the slower one-at-a-time route when running on .NET Core.

> but the above code makes me think the multi-targeting is not setup correctly

Can you explain what you mean here? Multi-targeting is set up correctly in this project. I followed the same pattern that we do in https://github.com/dotnet/runtime, which is to not use `#if` but instead split the different code into separate files, which makes it easier to reason about which code runs where. And it also aids in maintainability.

Reply:
What is an example of an API in NetCore that is not in NetStandard 2.0 or 2.1 ?

Maybe I am taking a purist approach - But i thought the generally accepted practice is to make libraries target NetStandard...   if you build an app you target NetCore or Framework and then implementation specific api's are use by the run-time.

Reply:
> What is an example of an API in NetCore that is not in NetStandard 2.0 or 2.1 ?

All the methods on Stream that accept Span / Memory parameters instead of arrays.

ex. https://github.com/apache/arrow/blob/249a12f131fa9e53aff035e5e5ad4ee6e38c3d05/csharp/src/Apache.Arrow/Extensions/StreamExtensions.netcoreapp2.1.cs#L24-L27

These methods are not in `netstandard2.0` and below. They were added in `netstandard2.1`, but the only executable frameworks that supports `netstandard2.1` is .NET Core 3.0 and 3.1. So if we don't target `netcoreapp2.1`, then someone running on .NET Core 2.1 will get the slower implementation.

> But i thought the generally accepted practice is to make libraries target NetStandard

That is correct, and this library DOES target netstandard. However, there are places where we want to take advantage of faster APIs (like the Stream.Read method above accepting a Span). In those situations, we compile multiple times - once for `netstandard` and once for the lowest framework that has the faster API (in this case `netcoreapp2.1`).

Consumers don't see anything different. They can still run on .NET Framework or Xamarin just fine. But when they run on .NET Core 2.1+, the code built specifically for that framework is used.

Reply:
Right, i get that.. but I am curious how often that is even the case.   I suppose when i started looking at Arrow NetCore 3.0/3.1 and netstandard 2.0 / 2.1 were available as options so my views are somewhat anachronistic? 

FWIW - i don't see too much performance difference with this patch applied - And I am dealing with 15 gig file generation.

Reply:
>  I suppose when i started looking at Arrow NetCore 3.0/3.1 and netstandard 2.0 / 2.1 were available as options so my views are somewhat anachronistic?

We don't expect apps to move up to new versions of .NET immediately. .NET Core 3.0 has only been out for 3-4 months, and 3.1 (the long term support version) just came out last month. So not many applications are on it yet. .NET Core 2.1 will be supported at a minimum until August 21, 2021 (see https://github.com/dotnet/core/blob/master/microsoft-support.md). So keeping it running as fast as possible on .NET Core 2.1 is important.

> i don't see too much performance difference with this patch applied - And I am dealing with 15 gig file generation.

The code change might not be on the hot path for your scenario, but that doesn't mean it isn't for some other scenario. For example, if someone was using `ArrowBuffer.Builder.Append(ReadOnlySpan<T>)` and `Build` on a hot path.

> i'll setup the net472 target in an update to the PR

I don't understand this. I wasn't asking for anything to do with `net472`.

Reply:
The patch I am supplying is to fix Span.CopyTo for a bug that only affects **.Net Framework** 

I thought creating a 3rd target for the arrow library (ie **net472**) is what you were proposing as a way to integrate this code so that only the target with the crash gets the fix - Or am I missing something?

Reply:
What I am proposing is to reuse the existing targets(`netstandard1.3` and `netcoreapp2.1`), and only doing the one-element-at-a-time copy on the `netstandard1.3` target, which is the one used when the consuming application is running on the .NET Framework. When an application is running on .NET Core (where there is no issue), it will use the Apache.Arrow assembly that is compiled for `netcoreapp2.1`. When compiling for `netcoreapp2.1`, we can use the built in `CopyTo` method.

Reply:
ok - i thought you were arguing for minimal surface area of the affected target frameworks (since netstandard1.3 is much wider range of run times.. mono, Xamarin, etc) and the bug seems to only be in net framework. However... I am happy to put into netstandard1.3 as I am currently using a fork of arrow and this bug is the only reason I am not using the official package.

what do you think of file globing for the separate files?  (see here:)
https://github.com/abbotware/abbotware/blob/master/Common/Abbotware.Core/Abbotware.Core.csproj


the only issue doing this I have seen

Reply:
> what do you think of file globing for the separate files? 

As long as the Solution Explorer in VS works correctly with it, I would be fine with that solution.

Reply:
do you see both files now? I only see this with the current .csproj
![image](https://user-images.githubusercontent.com/2100564/71927047-a1d70200-3162-11ea-9cc0-b01eadb54242.png)

Reply:
in either case - review the PR now as I did the split and added the globbing